src/category_theory/abelian/exact.lean

```diff
@@ -304,6 +309,91 @@ instance reflects_exact_sequences_of_preserves_zero_morphisms_of_faithful (F : C
     rw [F.map_comp, ← hk, ← hl, category.assoc, reassoc_of hfg.2, zero_comp, comp_zero]
   end }
 
+end
+
+section
+variables {D : Type u₂} [category.{v₁} D] [abelian D]  -- TODO: Generalize universe level
+variables (F : C ⥤ D) [preserves_zero_morphisms F]
+
+/--
+A functor from preadditive category to an abelian category which preserves kernels,
+preserves arbitrary products.
+-/
+def is_limit_map_cone_binary_fan_of_preserves_kernels {X Y Z : C} (π₁ : Z ⟶ X) (π₂ : Z ⟶ Y)
+  [preserves_limit (parallel_pair π₂ 0) F] (i : is_limit (binary_fan.mk π₁ π₂)) :
+  is_limit (F.map_cone (binary_fan.mk π₁ π₂)) :=
+let bc := binary_bicone.of_limit_cone i in
+let ibc := bicone_is_bilimit_of_limit_cone_of_is_limit i in
+begin
+  haveI : preserves_limit (parallel_pair bc.snd 0) F, { simpa },
+  let i_f : is_limit bc.snd_kernel_fork := binary_bicone.is_limit_snd_kernel_fork i,
+  have hex : exact (F.map bc.inl) (F.map bc.snd),
+  { fapply abelian.exact_of_is_kernel,
+    { rw [←F.map_comp, bc.inl_snd, F.map_zero] },
+    { exact (is_limit_map_cone_fork_equiv' F _).to_fun (is_limit_of_preserves F i_f) } },
+  exact (is_limit_map_cone_binary_fan_equiv F π₁ π₂).inv_fun
+    (is_bilimit_binary_bicone_of_split_mono_of_cokernel
+      (abelian.is_colimit_of_exact_of_epi (F.map bc.inl) (F.map bc.snd) hex)).is_limit
+end
+
+instance preserves_binary_product_of_preserves_kernels
+  [pres_kernels : ∀ {X Y} (f : X ⟶ Y), preserves_limit (parallel_pair f 0) F] {X Y : C} :
+  preserves_limit (pair X Y) F :=
+{ preserves := λ c,
+  begin
+    have : c = binary_fan.mk (binary_fan.fst c) (binary_fan.snd c),
+    { cases c, simp only [binary_fan.mk, const.obj_obj, eq_self_iff_true, heq_iff_eq, true_and],
+      ext, cases x; refl },
+    rw this,
+    apply is_limit_map_cone_binary_fan_of_preserves_kernels F
+  end }
```

Comment:
I'm still not convinced that this is the right approach, especially since we're defining data here. I'd much rather have this:
```suggestion
def iso_binary_fan_mk {X Y : C} (c : binary_fan X Y) : c ≅ binary_fan.mk c.fst c.snd :=
cones.ext (iso.refl _) (λ j, by cases j; tidy)

instance preserves_binary_product_of_preserves_kernels
  [pres_kernels : ∀ {X Y} (f : X ⟶ Y), preserves_limit (parallel_pair f 0) F] {X Y : C} :
  preserves_limit (pair X Y) F :=
  { preserves := λ c hc, limits.is_limit.of_iso_limit
    (is_limit_map_cone_binary_fan_of_preserves_kernels F _ _
      (is_limit.of_iso_limit hc (iso_binary_fan_mk c)))
    ((cones.functoriality _ F).map_iso (iso_binary_fan_mk c).symm) }
```

With #14030, this can be streamlined slightly.

Reply:
I agree that it's not using the equality proof here, so the choice is between `iso_binary_fan_mk` and a custom recursor for `binary_fan`. Both internally (ab)use the same equality/extensionality of cones, and conceal it from the user, but I'm if the iso lemma is the current standard in mathlib, let's do it that way, maybe I'll open a topic on Zulip... Isn't the `simp` behaviour better if we don't package it in an iso?

Reply:
A Zulip thread sounds good.